### PR TITLE
random: add a switch of using rdrand ins in _extract_crng

### PR DESCRIPTION
**Pull request by mengensun**
on some old AMD CPU(eg: older then AMD EPYC 7K62) , rdrand is slowly,
random using rdrand to make the random num more hard to be guessed.
while the randomness is not really affect by rdrand.

so, when user not using random for security case, eg: just geting some
random num, not using it to generate password, using rdrand on old
amd process is so expensive.

so, here add a swith to rdrand, by default using the rdrand, when
closed using tsc, jiffies, and entropy from the fast_pool to emulate
the rdrand

testing on AMD EPYC 7K62 like follow:

---

echo 1 > /proc/sys/kernel/random/tos_use_rdrand
stress-ng --urandom 2 --timeout 10s --times --metrics
bogo ops real time  usr time  sys time   bogo ops/s   bogo ops/s
           (secs)    (secs)    (secs)   (real time) (usr+sys time)
   16542     10.00      0.00     19.98      1654.09       827.93

---

echo 0 > /proc/sys/kernel/random/tos_use_rdrand
stress-ng --urandom 2 --timeout 10s --times --metrics
bogo ops real time  usr time  sys time   bogo ops/s   bogo ops/s
           (secs)    (secs)    (secs)   (real time) (usr+sys time)
  222005     10.00      0.05     19.93     22200.41     11111.36

/dev/urandom reading and geturandom  syscall 13 times faster then using
really rdrand.

Signed-off-by:mengensun <mengensun@tencent.com>
Reviewed-by: supowang<supowang@tencent.com>
Reviewed-by: samuelliao<samuelliao@tencent.com>
Reviewed-by: Ruippan<Ruippan@tencent.com>

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/TencentOS-kernel?pullRequest=201) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/TencentOS-kernel?pullRequest=201) before we can accept your contribution.<br/><hr/>**mengensun** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/TencentOS-kernel?pullRequest=201) it.</sub>

File: drivers/char/random.c [hunk 1]
```diff
@@ -1057,9 +1061,15 @@ static void _extract_crng(struct crng_state *crng,
...
 	     time_after(jiffies, crng->init_time + CRNG_RESEED_INTERVAL)))
 		crng_reseed(crng, crng == &primary_crng ? &input_pool : NULL);
+
 	spin_lock_irqsave(&crng->lock, flags);
-	if (arch_get_random_long(&v))
+	/* primary_crng is seldomly extracted */
+	if (!use_rdrand && crng != &primary_crng) {
+		v = simulation_rdrand();
+		crng->state[14] ^= v;
+	} else if (arch_get_random_long(&v))
```
Comment (frankjpliu) [inline drivers/char/random.c:1070]: else if 分支添加大括号。

File: drivers/char/random.c [hunk 2]
```diff
@@ -1268,6 +1278,21 @@ void add_input_randomness(unsigned int type, unsigned int code,
 EXPORT_SYMBOL_GPL(add_input_randomness);
 
 static DEFINE_PER_CPU(struct fast_pool, irq_randomness);
+static unsigned long simulation_rdrand()
```
Comment (frankjpliu) [inline drivers/char/random.c:1281]: 找 tallwu 或者 anakinzhang 确认下，这种特性，是否会影响安全认证。

File: drivers/char/random.c (hunk 1 above)
Comment (mengensun) [reply drivers/char/random.c:1070]: 好的，谢谢，我改改

File: drivers/char/random.c (hunk 2 above)
Comment (mengensun) [reply drivers/char/random.c:1281]: 好的，感谢

Comment (tallwu) [reply drivers/char/random.c:1281]: 随机数这里不影响安全认证。

**Pull request merged**